### PR TITLE
Add Git tags support

### PR DESCRIPTION
**Pull request by hashimoon**
## Description

Adding support to Git Tags 

## Testing plan

Use resource like the following:

```tf
locals {
  organization = "..."
  oauth_token_id = "..."
  identifier = "user/repo
}

resource "tfe_workspace" "my-workspace" {
  name                  = "my-workspace"
  organization          = local.organization
  allow_destroy_plan    = true
  auto_apply            = true
  execution_mode        = "remote"
  speculative_enabled   = true
  file_triggers_enabled = false

  vcs_repo {
    identifier         = local.identifier
    ingress_submodules = false
    oauth_token_id = local.oauth_token_id
    tags_regex = "^\\d+.\\d+.\\d+$" # semantic versioning
  }
}
```

1. Verify the correct settings were set in TF Cloud/TFE VCS settings for the workspace that git tags semantic versioning is selected
   * Release a new tagged release in the repo specified in `local.identifier`
   * See a new run has been trigged
1. Change configuration to the following

```tf
locals {
  organization = "..." # set
  oauth_token_id = "..." # set
  identifier = "user/repo
  working_directory = "..." # set
}

resource "tfe_workspace" "my-workspace" {
  name                  = "my-workspace"
  organization          = local.organization
  allow_destroy_plan    = true
  auto_apply            = true
  execution_mode        = "remote"
  speculative_enabled   = true
  file_triggers_enabled = true
  working_directory = local.working_directory

  vcs_repo {
    identifier         = local.identifier
    ingress_submodules = false
    oauth_token_id = local.oauth_token_id
  }
}
```

1. Terraform plan/apply succeeds
1. Verify in TF Cloud/TFE in VCS settings git tags is not selected.

## External links

_Include any links here that might be helpful for people reviewing your PR. If there are none, feel free to delete this section._

- [go-tfe documentation](https://pkg.go.dev/github.com/hashicorp/go-tfe#VCSRepo)
- [API documentation PR](https://github.com/hashicorp/terraform-docs-common/pull/64)
- [Related PR](https://github.com/hashicorp/terraform-provider-tfe/pull/xxxx)

## Output from acceptance tests

_Please run applicable acceptance tests locally and include the output here. See TESTS.md to learn how to run acceptance tests._

_If you are an external contributor, your contribution(s) will first be reviewed before running them against the project's CI pipeline._

```
$ TESTARGS="-run TestAccTFEWorkspace" make testacc

...
```

### CODE REVIEW

File: tfe/data_source_workspace.go [hunk 1]
```diff
@@ -231,6 +236,7 @@ func dataSourceTFEWorkspaceRead(d *schema.ResourceData, meta interface{}) error
 			"branch":             workspace.VCSRepo.Branch,
 			"ingress_submodules": workspace.VCSRepo.IngressSubmodules,
 			"oauth_token_id":     workspace.VCSRepo.OAuthTokenID,
+			"tags_regex":         workspace.VCSRepo.TagsRegex,
```
Comment (omarismail) [inline tfe/data_source_workspace.go:239]: Could we update the [tests](https://github.com/hashicorp/terraform-provider-tfe/blob/main/tfe/data_source_workspace_test.go) for this.

File: tfe/resource_tfe_policy_set_test.go [hunk 2]
```diff
@@ -997,6 +1005,7 @@ resource "tfe_policy_set" "foobar" {
     branch             = "foo"
     ingress_submodules = true
     oauth_token_id     = "id"
+    tags_regex 		   = ""
   }
 } `, rInt, sourcePath)
 }
```
Comment (omarismail) [inline tfe/resource_tfe_policy_set_test.go:1011]: Can we add a test where we set the tags_regex?

File: tfe/resource_tfe_workspace_test.go [hunk 3]
```diff
@@ -860,6 +864,8 @@ func TestAccTFEWorkspace_importVCSBranch(t *testing.T) {
 						"tfe_workspace.foobar", "vcs_repo.0.branch", GITHUB_WORKSPACE_BRANCH),
 					resource.TestCheckResourceAttr(
 						"tfe_workspace.foobar", "vcs_repo.0.ingress_submodules", "false"),
+					resource.TestCheckResourceAttr(
+						"tfe_workspace.foobar", "vcs_repo.0.tags_regex", ""),
```
Comment (omarismail) [inline tfe/resource_tfe_workspace_test.go:868]: Likewise as above test comment. Could we add a test where we set tags_regex, and another test where we update it and delete it? Going through the full CRUD lifecycle.

**Review by omarismail — CHANGES_REQUESTED** [review]
Overall looks good, but we should add a number of tests for this

File: tfe/resource_tfe_workspace.go [hunk 4]
```diff
@@ -211,6 +211,11 @@ func resourceTFEWorkspace() *schema.Resource {
 							Type:     schema.TypeString,
 							Required: true,
 						},
+
+						"tags_regex": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Comment (omarismail) [inline tfe/resource_tfe_workspace.go:217]: We'll need to add a `ConflictsWith: []string{"trigger_prefixes", "trigger_prefixes", "file_triggers_enabled"}` to keep this in sync with the API

File: tfe/resource_tfe_workspace_migrate.go [hunk 5]
```diff
@@ -95,6 +95,11 @@ func resourceTfeWorkspaceResourceV0() *schema.Resource {
 							Type:     schema.TypeString,
 							Required: true,
 						},
+
+						"tags_regex": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
```
Comment (omarismail) [inline tfe/resource_tfe_workspace_migrate.go:102]: likewise here with the `ConflictsWith`

**Review by omarismail — CHANGES_REQUESTED** [review]
great addition of tests, just one more change needed to use `ConflictsWith` to keep consistent with the API

**Review by omarismail — APPROVED** [review]
lgtm! I think there are some test failures tho

**Comment by hashimoon** [comment]
> great addition of tests, just one more change needed to use `ConflictsWith` to keep consistent with the API

Is `ConflictsWith` going to be applicable if we need to set the other values as `nil`/`[]`?

**Comment by brandonc** [comment]
@omarismail @hashimoon Do you think it would be best to land this after the beta flag is removed and the feature is GA?

File: tfe/resource_tfe_workspace_migrate.go [hunk 6]
```diff
@@ -95,6 +95,12 @@ func resourceTfeWorkspaceResourceV0() *schema.Resource {
 							Type:     schema.TypeString,
 							Required: true,
 						},
+
+						"tags_regex": {
```
Comment (matejrisek) [inline tfe/resource_tfe_workspace_migrate.go:99]: learning: What do we actually migrate here? I would argue that v0 of workspace didn't have `tags_regex`. I might be wrong about the purpose of migrations in the tfe provider.

Comment (omarismail) [reply tfe/resource_tfe_workspace_migrate.go:99]: Yeah I don't think we need this. This is called from [here](https://github.com/hashicorp/terraform-provider-tfe/blob/d55c76c7e4d839b5712a125adcec84c5a8557a9e/tfe/resource_tfe_workspace.go#L28) and its [about state migration](https://www.terraform.io/plugin/sdkv2/resources/state-migration). We don't have tags_regex for in state, so no need to migrate anything.

File: tfe/resource_tfe_workspace_test.go [hunk 7]
```diff
@@ -759,6 +761,66 @@ func TestAccTFEWorkspace_updateVCSRepo(t *testing.T) {
...
+						"tfe_workspace.foobar", "vcs_repo.0.branch", GITHUB_WORKSPACE_BRANCH),
+					resource.TestCheckResourceAttr(
+						"tfe_workspace.foobar", "vcs_repo.0.ingress_submodules", "false"),
+					resource.TestCheckResourceAttr(
+						"tfe_workspace.foobar", "vcs_repo.0.tags_regex", ""),
+				),
+			},
+		},
+	})
+}
```
Comment (omarismail) [inline tfe/resource_tfe_workspace_test.go:822]: We should add another test where we  start a new config without `tags_regex`, then add it, then delete it

File: website/docs/r/workspace.html.markdown [hunk 8]
```diff
@@ -124,6 +124,7 @@ The `vcs_repo` block supports:
   cloning the VCS repository. Defaults to `false`.
 * `oauth_token_id` - (Required) The VCS Connection (OAuth Connection + Token) to use.
   This ID can be obtained from a `tfe_oauth_client` resource.
+* `tags_regex` - (Optional) A regular expression used to trigger a Workspace run for matching Git tags.
```
Comment (omarismail) [inline website/docs/r/workspace.html.markdown:127]: We should add that this conflicts with `trigger_patterns` and `trigger_prefixes`

**Review by omarismail — CHANGES_REQUESTED** [review]
A couple comments regarding:
1. Adding another test
2. removing the migration
3. updating docs

File: tfe/resource_tfe_workspace_test.go [hunk 9]
```diff
@@ -759,6 +761,66 @@ func TestAccTFEWorkspace_updateVCSRepo(t *testing.T) {
...
 }
 
+func TestAccTFEWorkspace_updateVCSRepoTagsRegex(t *testing.T) {
+	workspace := &tfe.Workspace{}
+	rInt := rand.New(rand.NewSource(time.Now().UnixNano())).Int()
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			if GITHUB_TOKEN == "" {
```
Comment (matejrisek) [inline tfe/resource_tfe_workspace_test.go:771]: This is third time these checks are repeated in this file, IMHO it warrants method extraction

**Comment by mjyocca** [comment]
Updating go-tfe package with [pr](https://github.com/hashicorp/go-tfe/pull/468) fixes the current broken test

**Review by omarismail — APPROVED** [review]
lgtm!

**Comment by brandonc** [comment]
I re-ran the tests in SSH so I could extract the GITHUB_WORKSPACE config and cancelling it caused the CI failure. I will straighten it out!

**Review by brandonc — APPROVED** [review]
Works well!

I had some trouble removing the entire vcs_repo config. This did not seem to delete the ingress trigger from the workspace, so tags_regex continues to trigger conflicts with trigger_prefixes etc. I think this problem is not specific to git_tags and probably is a separate provider bug.

**Pull request merged**